### ISSUE TITLE
Dedicated etcd nodes

### ISSUE THREAD

**Issue by mumoshu** [kubernetes/kubeadm#491]
This is probably a **FEATURE REQUEST**.

---

Would it be possible to add support for dedicated etcd nodes into kubeadm?

As of today, several k8s cluster provisioners like kops, kube-aws, etc. supports the setup uses dedicated nodes for etcd - in other words, etcd nodes are separated from master/control-plane/controller nodes running apiserver, controller-manager, and so on - for extra reliability of etcd clusters for large k8s clusters.
For me, the setup seems to make sense regardless of we use kubeadm or not for cluster bootstrapping.

You may already know but in case you missed them, please see https://github.com/coreos/etcd-operator/issues/40 and https://coreos.com/operators/etcd/docs/latest/best_practices.html for more explanations of the setup.

Some more clarifications:
* This feature request is different from (re)using external etcd clusters by setting `etcd.endpoints` in a kubeadm master confguration file like described in https://kubernetes.io/docs/admin/kubeadm/#sample-master-configuration.
* Even though kubeadm doesn't support self-hosted etcd today as mentioned in https://kubernetes.io/docs/admin/kubeadm/#caveats, I believe it doesn't affect this feature request. I'd rather want kubeadm to support dedicated etcd nodes regardless of etcd is self-hosted or not.
* In case this feature request could be addressed, the timing would be after #261, at least.

**Comment by luxas** [comment]
Provisioning etcd clusters is out of scope for kubeadm. You can provision a single-node etcd instance listening on localhost with `kubeadm alpha phase etcd local`, but we don't want to go down the route doing something more fancy just yet as there are a bazillion of ways you can set up a cluster, and that is different from provider to provider.

We do support pointing to an external etcd cluster though. I think that should be enough for your use-case. Does that make sense? Thanks!

**Comment by jamiehannaford** [comment]
Agree with @luxas. Since kubeadm deploys etcd using kubernetes, the kubelet would need to be installed on the etcd node as a minimum, which seems a bit sub-optimal IMO. We don't want to go down the path of making kubeadm deploy distro-specific configuration files like systemd either. 

I would like kubeadm to support self-hosted etcd at some point, though.

**Comment by luxas** [comment]
@mumoshu I'm closing this, as it is out of scope to create generic etcd clusters. We do what's needed to get a master up and running, but won't reinvent the wheel for generic clusters, there are many other good tools for that

**Comment by mumoshu** [comment]
@luxas @jamiehannaford Thank you very much for the detailed explanations and clarifications - I now agree that kubeadm should not re-invent the wheel here.

> I would like kubeadm to support self-hosted etcd at some point, though.

Just curious but even after that, would the dedicated etcd node(s) support be out-of-scope?

Thanks for maintaining the great project btw! Really looking forward to utilize kubeadm in my work.

**Comment by luxas** [comment]
> Just curious but even after that, would the dedicated etcd node(s) support be out-of-scope?

Yes, dedicated etcd nodes would be out of scope still. However, there would be one etcd peer per master node, colocated with the API server

**Comment by mumoshu** [comment]
@luxas I see - thanks again for the clarification 👍

**Comment by nelsonfassis** [comment]
how can I specify the external etcd cluster? Didnt find in the documentation the option for that. Do I must use an yaml file for that?

**Comment by jamiehannaford** [comment]
@nelsonfassis If you look at the docs for the config file (https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#config-file), you can specify them as a YAML list. See etcd > endpoints section.

**Comment by nelsonfassis** [comment]
@jamiehannaford exactly, I'd like to use something like kubeadm init --etcd-cluster="<cluster>". So I guess it is not an option, right?

**Comment by jamiehannaford** [comment]
@nelsonfassis That CLI flag is not supported, you will need to use the configuration file.

**Comment by nelsonfassis** [comment]
@jamiehannaford Thank you for the clarification. I set up a cluster without making the etcd cluster external of kubernetes, so now I have only one etcd pod running which is a very dangerous setup.
1 - Is it safe to create an etcd cluster on the hosts (same as my k8s cluster, but running on the server instead of containers), add those nodes to the single k8s etcd and then remove etcd pod?

2 - What would be the implication of running instances of this etcd pod on all nodes as Daemonsets? Wouldn't it be more reliable and as easy to configure with kubeadm as the current way?

Thank you for your help :)

**Comment by jamiehannaford** [comment]
@nelsonfassis It depends on your use case. If you're using k8s for informal workloads (dev, staging, nonpublic) then self-hosted etcd via kubeadm is usually fine. For production workloads we recommend something a bit more robust like systemd units, either on separate servers or colocated on the masters.

**Comment by dylanfoster** [comment]
I've specified my external etcd cluster in a basic config.yml file but my kube cluster fails to fully boot. I'm running this in a boot script for an aws instance. `kubeadm init` succeeds however cni fails to connect it seems using weave.

My basic steps are

- `kubeadm init`
- `sysctl net.bridge.bridge-nf-call-iptables=1`
- `kubectl apply -f <weave>`

config.yml
```yml
apiVersion: kubeadm.k8s.io/v1alpha1
kind: MasterConfiguration
etcd:
  endpoints:
    - http://${etcd_1_ip}:2379
    - http://${etcd_2_ip}:2379
    - http://${etcd_3_ip}:2379
token: ${k8stoken}
```

kubeadm init --config config.yml

journalctl output: (truncated)

```
Mar 23 03:32:45 ip-10-0-10-187 kubelet[5992]: I0323 03:32:45.416696    5992 kuberuntime_manager.go:758] checking backoff for container "weave" in pod "weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:32:45 ip-10-0-10-187 kubelet[5992]: I0323 03:32:45.416803    5992 kuberuntime_manager.go:768] Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)
Mar 23 03:32:45 ip-10-0-10-187 kubelet[5992]: E0323 03:32:45.416830    5992 pod_workers.go:186] Error syncing pod 42c43015-2e4a-11e8-855e-062e2455e2a8 ("weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"), skipping: failed to "StartContainer" for "weave" with CrashLoopBackOff: "Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:32:49 ip-10-0-10-187 kubelet[5992]: W0323 03:32:49.278343    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:32:49 ip-10-0-10-187 kubelet[5992]: E0323 03:32:49.278619    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:32:54 ip-10-0-10-187 kubelet[5992]: W0323 03:32:54.279481    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:32:54 ip-10-0-10-187 kubelet[5992]: E0323 03:32:54.279594    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:32:59 ip-10-0-10-187 kubelet[5992]: W0323 03:32:59.280496    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:32:59 ip-10-0-10-187 kubelet[5992]: E0323 03:32:59.280638    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:32:59 ip-10-0-10-187 kubelet[5992]: I0323 03:32:59.375606    5992 kuberuntime_manager.go:514] Container {Name:weave Image:weaveworks/weave-kube:2.2.1 Command:[/home/weave/launch.sh] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[{Name:HOSTNAME Value: ValueFrom:&EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:spec.nodeName,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] Resources:{Limits:map[] Requests:map[cpu:{i:{value:10 scale:-3} d:{Dec:<nil>} s:10m Format:DecimalSI}]} VolumeMounts:[{Name:weavedb ReadOnly:false MountPath:/weavedb SubPath: MountPropagation:<nil>} {Name:cni-bin ReadOnly:false MountPath:/host/opt SubPath: MountPropagation:<nil>} {Name:cni-bin2 ReadOnly:false MountPath:/host/home SubPath: MountPropagation:<nil>} {Name:cni-conf ReadOnly:false MountPath:/host/etc SubPath: MountPropagation:<nil>} {Name:dbus ReadOnly:false MountPath:/host/var/lib/dbus SubPath: MountPropagation:<nil>} {Name:lib-modules ReadOnly:false MountPath:/lib/modules SubPath: MountPropagation:<nil>} {Name:xtables-lock ReadOnly:false MountPath:/run/xtables.lock SubPath: MountPropagation:<nil>} {Name:weave-net-token-k497l ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath: MountPropagation:<nil>}] VolumeDevices:[] LivenessProbe:&Probe{Handler:Handler{Exec:nil,HTTPGet:&HTTPGetAction{Path:/status,Port:6784,Host:127.0.0.1,Scheme:HTTP,HTTPHeaders:[],},TCPSocket:nil,},InitialDelaySeconds:30,TimeoutSeconds:1,PeriodSeconds:10,SuccessThreshold:1,FailureThreshold:3,} ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:IfNotPresent SecurityContext:&SecurityContext{Capabilities:nil,Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,} Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
Mar 23 03:32:59 ip-10-0-10-187 kubelet[5992]: I0323 03:32:59.375801    5992 kuberuntime_manager.go:758] checking backoff for container "weave" in pod "weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:32:59 ip-10-0-10-187 kubelet[5992]: I0323 03:32:59.375931    5992 kuberuntime_manager.go:768] Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)
Mar 23 03:32:59 ip-10-0-10-187 kubelet[5992]: E0323 03:32:59.375975    5992 pod_workers.go:186] Error syncing pod 42c43015-2e4a-11e8-855e-062e2455e2a8 ("weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"), skipping: failed to "StartContainer" for "weave" with CrashLoopBackOff: "Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:33:04 ip-10-0-10-187 kubelet[5992]: W0323 03:33:04.281504    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:33:04 ip-10-0-10-187 kubelet[5992]: E0323 03:33:04.281768    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:33:09 ip-10-0-10-187 kubelet[5992]: W0323 03:33:09.282705    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:33:09 ip-10-0-10-187 kubelet[5992]: E0323 03:33:09.283072    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:33:12 ip-10-0-10-187 kubelet[5992]: I0323 03:33:12.377154    5992 kuberuntime_manager.go:514] Container {Name:weave Image:weaveworks/weave-kube:2.2.1 Command:[/home/weave/launch.sh] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[{Name:HOSTNAME Value: ValueFrom:&EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:spec.nodeName,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] Resources:{Limits:map[] Requests:map[cpu:{i:{value:10 scale:-3} d:{Dec:<nil>} s:10m Format:DecimalSI}]} VolumeMounts:[{Name:weavedb ReadOnly:false MountPath:/weavedb SubPath: MountPropagation:<nil>} {Name:cni-bin ReadOnly:false MountPath:/host/opt SubPath: MountPropagation:<nil>} {Name:cni-bin2 ReadOnly:false MountPath:/host/home SubPath: MountPropagation:<nil>} {Name:cni-conf ReadOnly:false MountPath:/host/etc SubPath: MountPropagation:<nil>} {Name:dbus ReadOnly:false MountPath:/host/var/lib/dbus SubPath: MountPropagation:<nil>} {Name:lib-modules ReadOnly:false MountPath:/lib/modules SubPath: MountPropagation:<nil>} {Name:xtables-lock ReadOnly:false MountPath:/run/xtables.lock SubPath: MountPropagation:<nil>} {Name:weave-net-token-k497l ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath: MountPropagation:<nil>}] VolumeDevices:[] LivenessProbe:&Probe{Handler:Handler{Exec:nil,HTTPGet:&HTTPGetAction{Path:/status,Port:6784,Host:127.0.0.1,Scheme:HTTP,HTTPHeaders:[],},TCPSocket:nil,},InitialDelaySeconds:30,TimeoutSeconds:1,PeriodSeconds:10,SuccessThreshold:1,FailureThreshold:3,} ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:IfNotPresent SecurityContext:&SecurityContext{Capabilities:nil,Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,} Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
Mar 23 03:33:12 ip-10-0-10-187 kubelet[5992]: I0323 03:33:12.377334    5992 kuberuntime_manager.go:758] checking backoff for container "weave" in pod "weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:33:12 ip-10-0-10-187 kubelet[5992]: I0323 03:33:12.377447    5992 kuberuntime_manager.go:768] Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)
Mar 23 03:33:12 ip-10-0-10-187 kubelet[5992]: E0323 03:33:12.377474    5992 pod_workers.go:186] Error syncing pod 42c43015-2e4a-11e8-855e-062e2455e2a8 ("weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"), skipping: failed to "StartContainer" for "weave" with CrashLoopBackOff: "Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:33:14 ip-10-0-10-187 kubelet[5992]: W0323 03:33:14.284256    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:33:14 ip-10-0-10-187 kubelet[5992]: E0323 03:33:14.284562    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:33:19 ip-10-0-10-187 kubelet[5992]: W0323 03:33:19.285474    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:33:19 ip-10-0-10-187 kubelet[5992]: E0323 03:33:19.285622    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:33:23 ip-10-0-10-187 kubelet[5992]: I0323 03:33:23.375630    5992 kuberuntime_manager.go:514] Container {Name:weave Image:weaveworks/weave-kube:2.2.1 Command:[/home/weave/launch.sh] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[{Name:HOSTNAME Value: ValueFrom:&EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:spec.nodeName,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] Resources:{Limits:map[] Requests:map[cpu:{i:{value:10 scale:-3} d:{Dec:<nil>} s:10m Format:DecimalSI}]} VolumeMounts:[{Name:weavedb ReadOnly:false MountPath:/weavedb SubPath: MountPropagation:<nil>} {Name:cni-bin ReadOnly:false MountPath:/host/opt SubPath: MountPropagation:<nil>} {Name:cni-bin2 ReadOnly:false MountPath:/host/home SubPath: MountPropagation:<nil>} {Name:cni-conf ReadOnly:false MountPath:/host/etc SubPath: MountPropagation:<nil>} {Name:dbus ReadOnly:false MountPath:/host/var/lib/dbus SubPath: MountPropagation:<nil>} {Name:lib-modules ReadOnly:false MountPath:/lib/modules SubPath: MountPropagation:<nil>} {Name:xtables-lock ReadOnly:false MountPath:/run/xtables.lock SubPath: MountPropagation:<nil>} {Name:weave-net-token-k497l ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath: MountPropagation:<nil>}] VolumeDevices:[] LivenessProbe:&Probe{Handler:Handler{Exec:nil,HTTPGet:&HTTPGetAction{Path:/status,Port:6784,Host:127.0.0.1,Scheme:HTTP,HTTPHeaders:[],},TCPSocket:nil,},InitialDelaySeconds:30,TimeoutSeconds:1,PeriodSeconds:10,SuccessThreshold:1,FailureThreshold:3,} ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:IfNotPresent SecurityContext:&SecurityContext{Capabilities:nil,Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,} Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
Mar 23 03:33:23 ip-10-0-10-187 kubelet[5992]: I0323 03:33:23.375784    5992 kuberuntime_manager.go:758] checking backoff for container "weave" in pod "weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:33:23 ip-10-0-10-187 kubelet[5992]: I0323 03:33:23.375893    5992 kuberuntime_manager.go:768] Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)
Mar 23 03:33:23 ip-10-0-10-187 kubelet[5992]: E0323 03:33:23.375920    5992 pod_workers.go:186] Error syncing pod 42c43015-2e4a-11e8-855e-062e2455e2a8 ("weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"), skipping: failed to "StartContainer" for "weave" with CrashLoopBackOff: "Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:33:24 ip-10-0-10-187 kubelet[5992]: W0323 03:33:24.286615    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:33:24 ip-10-0-10-187 kubelet[5992]: E0323 03:33:24.287001    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:33:29 ip-10-0-10-187 kubelet[5992]: W0323 03:33:29.287944    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:33:29 ip-10-0-10-187 kubelet[5992]: E0323 03:33:29.288063    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:33:34 ip-10-0-10-187 kubelet[5992]: W0323 03:33:34.288988    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:33:34 ip-10-0-10-187 kubelet[5992]: E0323 03:33:34.289133    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:33:36 ip-10-0-10-187 kubelet[5992]: I0323 03:33:36.375698    5992 kuberuntime_manager.go:514] Container {Name:weave Image:weaveworks/weave-kube:2.2.1 Command:[/home/weave/launch.sh] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[{Name:HOSTNAME Value: ValueFrom:&EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:spec.nodeName,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] Resources:{Limits:map[] Requests:map[cpu:{i:{value:10 scale:-3} d:{Dec:<nil>} s:10m Format:DecimalSI}]} VolumeMounts:[{Name:weavedb ReadOnly:false MountPath:/weavedb SubPath: MountPropagation:<nil>} {Name:cni-bin ReadOnly:false MountPath:/host/opt SubPath: MountPropagation:<nil>} {Name:cni-bin2 ReadOnly:false MountPath:/host/home SubPath: MountPropagation:<nil>} {Name:cni-conf ReadOnly:false MountPath:/host/etc SubPath: MountPropagation:<nil>} {Name:dbus ReadOnly:false MountPath:/host/var/lib/dbus SubPath: MountPropagation:<nil>} {Name:lib-modules ReadOnly:false MountPath:/lib/modules SubPath: MountPropagation:<nil>} {Name:xtables-lock ReadOnly:false MountPath:/run/xtables.lock SubPath: MountPropagation:<nil>} {Name:weave-net-token-k497l ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath: MountPropagation:<nil>}] VolumeDevices:[] LivenessProbe:&Probe{Handler:Handler{Exec:nil,HTTPGet:&HTTPGetAction{Path:/status,Port:6784,Host:127.0.0.1,Scheme:HTTP,HTTPHeaders:[],},TCPSocket:nil,},InitialDelaySeconds:30,TimeoutSeconds:1,PeriodSeconds:10,SuccessThreshold:1,FailureThreshold:3,} ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:IfNotPresent SecurityContext:&SecurityContext{Capabilities:nil,Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,} Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
Mar 23 03:33:36 ip-10-0-10-187 kubelet[5992]: I0323 03:33:36.375869    5992 kuberuntime_manager.go:758] checking backoff for container "weave" in pod "weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:33:36 ip-10-0-10-187 kubelet[5992]: I0323 03:33:36.375973    5992 kuberuntime_manager.go:768] Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)
Mar 23 03:33:36 ip-10-0-10-187 kubelet[5992]: E0323 03:33:36.376012    5992 pod_workers.go:186] Error syncing pod 42c43015-2e4a-11e8-855e-062e2455e2a8 ("weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"), skipping: failed to "StartContainer" for "weave" with CrashLoopBackOff: "Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:33:39 ip-10-0-10-187 kubelet[5992]: W0323 03:33:39.290080    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:33:39 ip-10-0-10-187 kubelet[5992]: E0323 03:33:39.290246    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:33:44 ip-10-0-10-187 kubelet[5992]: W0323 03:33:44.291285    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:33:44 ip-10-0-10-187 kubelet[5992]: E0323 03:33:44.291723    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:33:49 ip-10-0-10-187 kubelet[5992]: W0323 03:33:49.292630    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:33:49 ip-10-0-10-187 kubelet[5992]: E0323 03:33:49.293025    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 23 03:33:52 ip-10-0-10-187 kubelet[5992]: I0323 03:33:52.375766    5992 kuberuntime_manager.go:514] Container {Name:weave Image:weaveworks/weave-kube:2.2.1 Command:[/home/weave/launch.sh] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[{Name:HOSTNAME Value: ValueFrom:&EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:spec.nodeName,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] Resources:{Limits:map[] Requests:map[cpu:{i:{value:10 scale:-3} d:{Dec:<nil>} s:10m Format:DecimalSI}]} VolumeMounts:[{Name:weavedb ReadOnly:false MountPath:/weavedb SubPath: MountPropagation:<nil>} {Name:cni-bin ReadOnly:false MountPath:/host/opt SubPath: MountPropagation:<nil>} {Name:cni-bin2 ReadOnly:false MountPath:/host/home SubPath: MountPropagation:<nil>} {Name:cni-conf ReadOnly:false MountPath:/host/etc SubPath: MountPropagation:<nil>} {Name:dbus ReadOnly:false MountPath:/host/var/lib/dbus SubPath: MountPropagation:<nil>} {Name:lib-modules ReadOnly:false MountPath:/lib/modules SubPath: MountPropagation:<nil>} {Name:xtables-lock ReadOnly:false MountPath:/run/xtables.lock SubPath: MountPropagation:<nil>} {Name:weave-net-token-k497l ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath: MountPropagation:<nil>}] VolumeDevices:[] LivenessProbe:&Probe{Handler:Handler{Exec:nil,HTTPGet:&HTTPGetAction{Path:/status,Port:6784,Host:127.0.0.1,Scheme:HTTP,HTTPHeaders:[],},TCPSocket:nil,},InitialDelaySeconds:30,TimeoutSeconds:1,PeriodSeconds:10,SuccessThreshold:1,FailureThreshold:3,} ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:IfNotPresent SecurityContext:&SecurityContext{Capabilities:nil,Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,} Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
Mar 23 03:33:52 ip-10-0-10-187 kubelet[5992]: I0323 03:33:52.375926    5992 kuberuntime_manager.go:758] checking backoff for container "weave" in pod "weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:33:52 ip-10-0-10-187 kubelet[5992]: I0323 03:33:52.376042    5992 kuberuntime_manager.go:768] Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)
Mar 23 03:33:52 ip-10-0-10-187 kubelet[5992]: E0323 03:33:52.376071    5992 pod_workers.go:186] Error syncing pod 42c43015-2e4a-11e8-855e-062e2455e2a8 ("weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"), skipping: failed to "StartContainer" for "weave" with CrashLoopBackOff: "Back-off 1m20s restarting failed container=weave pod=weave-net-m4jx4_kube-system(42c43015-2e4a-11e8-855e-062e2455e2a8)"
Mar 23 03:33:54 ip-10-0-10-187 kubelet[5992]: W0323 03:33:54.294095    5992 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 23 03:33:54 ip-10-0-10-187 kubelet[5992]: E0323 03:33:54.294365    5992 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
```

**Comment by rackeric** [comment]
I am seeing the same issue as @dylanfoster, why are the logs complaining about no CNI found when that is listed as the next step in setting up the cluster?  Do we need to populate that entire configure file template or will kubeadm set all other values itself?
